Audit: type-drift between backend Settings allowlist and frontend Settings interface; stale TS errors in route tests

Split-off from the PR #543 self-review (part of #516 series). Two codebase-quality items that are worth tracking but were out of scope for any individual #516 PR.

## Item 1 — `Settings` interface type-drift

The pipeline-side allowlist in `apps/pipeline/app/services/notification_settings.py::_FIELDS` and the web-side TypeScript shape in `apps/web/src/components/NotificationSettingsSections.tsx::Settings` must stay in lockstep for settings round-tripping to work. Today this is enforced by code review only — neither side compiles against the other, so a new field added to one can silently drift.

We've added three batches of settings fields so far (inference, embedding, diarization) and the sync has held, but the blast radius of a typo here is real — the API proxy (`/api/notifications/settings/route.ts`) forwards arbitrary JSON, so bad TypeScript on the frontend could corrupt the DB.

**Ideas (any one would close this):**

1. **Runtime shape validation.** Add a Zod schema for `Settings` on the web side; parse the fetch response through it. Divergence would surface as a runtime error on the Settings page rather than a silent prod corruption.
2. **Codegen from the Python allowlist.** A small script that reads `_FIELDS` + `_SENSITIVE_FIELDS` + `_VALID_*` enums and emits the TypeScript interface. Wire it into `make deps-outdated`-style checks.
3. **Backend-owned types.** Ship an OpenAPI doc from FastAPI's existing auto-generation and consume it via `openapi-typescript` or similar. Same idea, more infrastructure.

I'd vote for option 1 (Zod) — cheapest to ship, catches the failure mode we care about.

## Item 2 — Pre-existing TS errors in test files

`npx tsc --noEmit` on `main` reports:

```
tests/unit/ask-coverage-route.test.ts(4,7): error TS2451: Cannot redeclare block-scoped variable 'mockQuery'.
tests/unit/ask-coverage-route.test.ts(7,6): error TS2300: Duplicate identifier 'RouteModule'.
tests/unit/ask-coverage-route.test.ts(8,5): error TS2451: Cannot redeclare block-scoped variable 'GET'.
tests/unit/queue-route.test.ts(10,7): error TS2451: Cannot redeclare block-scoped variable 'mockQuery'.
tests/unit/queue-route.test.ts(13,6): error TS2300: Duplicate identifier 'RouteModule'.
tests/unit/queue-route.test.ts(14,5): error TS2451: Cannot redeclare block-scoped variable 'GET'.
tests/unit/queue-route.test.ts(17,9): error TS2322: Type ... is not assignable to type ...
```

Both files declare `const mockQuery = ...` at module top level. TypeScript's `**/*.ts` glob treats the `tests/` directory as one compilation unit, so the two files collide.

Verified present on `main` before any #516 work (`git stash && checkout main && npx tsc --noEmit`), so this was likely introduced earlier — jest itself doesn't care (each test file is its own module at runtime), only the full-project tsc run sees it.

**Fix:** either rename the shared helpers inside each file (`askCoverageMockQuery`, `queueMockQuery`), or add `tests/**` to the default `tsconfig.json` `exclude` and rely on `tsconfig.test.json` alone for test-file type checking. The second option matches how many projects separate app vs. test compilation.

Acceptance: `npx tsc --noEmit` exits 0 on a clean `main` checkout.

## Scope

Not blocking any current feature. Item 1 is a design choice; item 2 is a 10-minute rename + tsconfig tweak. Pick up whenever.

--- a/apps/web/package.json
+++ b/apps/web/package.json
@@ -41,7 +41,8 @@
     "rehype-raw": "^7.0.0",
     "remark-gfm": "^4.0.1",
     "tailwind-merge": "^3.5.0",
-    "tailwindcss-animate": "^1.0.7"
+    "tailwindcss-animate": "^1.0.7",
+    "zod": "^4.3.6"
   },
   "devDependencies": {
     "@playwright/test": "^1.59.1",

--- a/apps/web/src/components/NotificationSettings.tsx
+++ b/apps/web/src/components/NotificationSettings.tsx
@@ -3,6 +3,7 @@
 import { useEffect, useState } from "react";
 import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";
 import { Settings, Toast } from "./NotificationSettingsSections";
+import { SettingsSchema } from "@/lib/settings-schema";
 import NotificationSection from "./NotificationSection";
 import RemoteInferenceSection from "./RemoteInferenceSection";
 
@@ -32,11 +33,26 @@ export default function NotificationSettings() {
   } | null>(null);
   const [dirtyNotifications, setDirtyNotifications] = useState<Partial<Settings>>({});
   const [dirtyInference, setDirtyInference] = useState<Partial<Settings>>({});
+  const [shapeError, setShapeError] = useState<string | null>(null);
 
   useEffect(() => {
     fetch("/api/notifications/settings")
       .then((r) => r.json())
-      .then((data) => setSettings(data))
+      .then((data) => {
+        const parsed = SettingsSchema.safeParse(data);
+        if (parsed.success) {
+          setSettings(parsed.data);
+          return;
+        }
+        // Drift between backend allowlist and frontend schema. Surface it
+        // visibly instead of writing back a corrupted shape on the next PUT.
+        const summary = parsed.error.issues
+          .slice(0, 5)
+          .map((iss) => `${iss.path.join(".") || "(root)"}: ${iss.message}`)
+          .join("; ");
+        console.error("settings_shape_mismatch", parsed.error.issues);
+        setShapeError(summary);
+      })
       .catch(() => {});
   }, []);
 
@@ -45,6 +61,24 @@ export default function NotificationSettings() {
     const id = setTimeout(() => setToast(null), 4000);
     return () => clearTimeout(id);
   }, [toast]);
+
+  if (shapeError) {
+    return (
+      <div className="border border-destructive/40 bg-destructive/10 rounded-md p-4 text-sm space-y-2">
+        <div className="font-medium text-destructive">
+          Settings response shape didn&apos;t match the expected schema.
+        </div>
+        <div className="text-muted-foreground">
+          The backend may have added or changed a field. Refusing to render
+          editable fields to avoid corrupting the database on save. Please
+          report this with the details below.
+        </div>
+        <pre className="text-xs bg-muted/40 rounded p-2 overflow-x-auto whitespace-pre-wrap">
+          {shapeError}
+        </pre>
+      </div>
+    );
+  }
 
   if (!settings) {
     return (

--- a/apps/web/src/components/NotificationSettingsSections.tsx
+++ b/apps/web/src/components/NotificationSettingsSections.tsx
@@ -1,39 +1,9 @@
 "use client";
 
-export interface Settings {
-  telegram_bot_token: string | null;
-  telegram_chat_id: string | null;
-  notification_email_to: string | null;
-  notification_email_from: string;
-  smtp_host: string;
-  smtp_port: number;
-  smtp_user: string | null;
-  smtp_password: string | null;
-  smtp_use_tls: boolean;
-  notification_frequency: string;
-  health_check_notifications_enabled: boolean;
-  inference_provider: "local" | "fireworks";
-  fireworks_api_key: string | null;
-  fireworks_audio_base_url: string;
-  fireworks_stt_model: string;
-  fireworks_stt_diarize: boolean;
-  fireworks_chat_base_url: string;
-  fireworks_chat_model: string;
-  fireworks_stt_cost_per_minute_usd: number;
-  embedding_provider: "local" | "fireworks";
-  embedding_model: string;
-  fireworks_embedding_base_url: string;
-  fireworks_embedding_model: string;
-  diarization_provider: "local" | "precision2";
-  pyannote_api_key: string | null;
-  pyannote_cloud_base_url: string;
-  pyannote_cloud_model: string;
-  pyannote_cloud_cost_per_second_usd: number;
-  telegram_configured: boolean;
-  email_configured: boolean;
-  fireworks_configured: boolean;
-  pyannote_cloud_configured: boolean;
-}
+// `Settings` is defined by the Zod schema in @/lib/settings-schema so the
+// runtime parse and the compile-time type cannot drift from each other.
+// Re-exported here to keep existing imports stable across the codebase.
+export type { Settings } from "@/lib/settings-schema";
 
 export function Toast({
   message,

--- a/apps/web/src/lib/settings-schema.ts
+++ b/apps/web/src/lib/settings-schema.ts
@@ -1,0 +1,67 @@
+import { z } from "zod";
+
+/**
+ * Runtime schema for the settings payload returned by
+ * `GET /api/notifications/settings`. The shape must stay in lockstep with
+ * the backend allowlist:
+ *
+ *   apps/pipeline/app/services/notification_settings.py::_FIELDS
+ *   (+ _NULLABLE_FIELDS, _VALID_FREQUENCIES, _VALID_INFERENCE_PROVIDERS,
+ *      _VALID_EMBEDDING_PROVIDERS, _VALID_DIARIZATION_PROVIDERS)
+ *
+ * The pipeline is the source of truth. If a new field is added there, mirror
+ * it here. Drift is caught at runtime when the Settings page parses the
+ * response — the page surfaces a clear error rather than silently using a
+ * mistyped shape (which previously could corrupt the DB on a follow-up PUT
+ * because /api/notifications/settings is a thin proxy that forwards arbitrary
+ * JSON to the backend).
+ */
+export const SettingsSchema = z.object({
+  // Nullable string fields (mirror _NULLABLE_FIELDS).
+  telegram_bot_token: z.string().nullable(),
+  telegram_chat_id: z.string().nullable(),
+  notification_email_to: z.string().nullable(),
+  smtp_user: z.string().nullable(),
+  smtp_password: z.string().nullable(),
+  fireworks_api_key: z.string().nullable(),
+  pyannote_api_key: z.string().nullable(),
+
+  // Required string fields.
+  notification_email_from: z.string(),
+  smtp_host: z.string(),
+  fireworks_audio_base_url: z.string(),
+  fireworks_stt_model: z.string(),
+  fireworks_chat_base_url: z.string(),
+  fireworks_chat_model: z.string(),
+  embedding_model: z.string(),
+  fireworks_embedding_base_url: z.string(),
+  fireworks_embedding_model: z.string(),
+  pyannote_cloud_base_url: z.string(),
+  pyannote_cloud_model: z.string(),
+
+  // Numbers.
+  smtp_port: z.number(),
+  fireworks_stt_cost_per_minute_usd: z.number(),
+  pyannote_cloud_cost_per_second_usd: z.number(),
+
+  // Booleans.
+  smtp_use_tls: z.boolean(),
+  health_check_notifications_enabled: z.boolean(),
+  fireworks_stt_diarize: z.boolean(),
+
+  // Enums (mirror backend _VALID_* sets).
+  notification_frequency: z.enum(["immediate", "daily", "weekly"]),
+  inference_provider: z.enum(["local", "fireworks"]),
+  embedding_provider: z.enum(["local", "fireworks"]),
+  diarization_provider: z.enum(["local", "precision2"]),
+
+  // Server-computed read-only flags. Not in _FIELDS but always present in
+  // the response payload — derived booleans the API includes alongside
+  // writable settings.
+  telegram_configured: z.boolean(),
+  email_configured: z.boolean(),
+  fireworks_configured: z.boolean(),
+  pyannote_cloud_configured: z.boolean(),
+});
+
+export type Settings = z.infer<typeof SettingsSchema>;

--- a/apps/web/tests/unit/settings-schema.test.ts
+++ b/apps/web/tests/unit/settings-schema.test.ts
@@ -1,0 +1,101 @@
+/**
+ * @jest-environment node
+ */
+import { SettingsSchema } from "@/lib/settings-schema";
+
+const validPayload = {
+  telegram_bot_token: null,
+  telegram_chat_id: null,
+  notification_email_to: null,
+  notification_email_from: "podlog@example.com",
+  smtp_host: "smtp.example.com",
+  smtp_port: 587,
+  smtp_user: null,
+  smtp_password: null,
+  smtp_use_tls: true,
+  notification_frequency: "daily",
+  health_check_notifications_enabled: true,
+  inference_provider: "local",
+  fireworks_api_key: null,
+  fireworks_audio_base_url: "https://api.fireworks.ai/inference/v1/audio",
+  fireworks_stt_model: "whisper-v3-turbo",
+  fireworks_stt_diarize: false,
+  fireworks_chat_base_url: "https://api.fireworks.ai/inference/v1",
+  fireworks_chat_model: "accounts/fireworks/models/llama-v3p3-70b-instruct",
+  fireworks_stt_cost_per_minute_usd: 0.0009,
+  embedding_provider: "local",
+  embedding_model: "all-MiniLM-L6-v2",
+  fireworks_embedding_base_url: "https://api.fireworks.ai/inference/v1",
+  fireworks_embedding_model: "nomic-ai/nomic-embed-text-v1.5",
+  diarization_provider: "local",
+  pyannote_api_key: null,
+  pyannote_cloud_base_url: "https://api.pyannote.ai/v1",
+  pyannote_cloud_model: "precision-2",
+  pyannote_cloud_cost_per_second_usd: 0.0006,
+  telegram_configured: false,
+  email_configured: false,
+  fireworks_configured: false,
+  pyannote_cloud_configured: false,
+};
+
+describe("SettingsSchema", () => {
+  it("accepts a known-good payload", () => {
+    const result = SettingsSchema.safeParse(validPayload);
+    expect(result.success).toBe(true);
+  });
+
+  it("rejects when a required field is missing", () => {
+    const { smtp_host: _omitted, ...rest } = validPayload;
+    const result = SettingsSchema.safeParse(rest);
+    expect(result.success).toBe(false);
+    if (!result.success) {
+      expect(result.error.issues.some((i) => i.path.includes("smtp_host"))).toBe(true);
+    }
+  });
+
+  it("rejects when a field has the wrong type (string instead of number)", () => {
+    const result = SettingsSchema.safeParse({ ...validPayload, smtp_port: "587" });
+    expect(result.success).toBe(false);
+    if (!result.success) {
+      expect(result.error.issues.some((i) => i.path.includes("smtp_port"))).toBe(true);
+    }
+  });
+
+  it("rejects an enum value outside the allowed set", () => {
+    const result = SettingsSchema.safeParse({ ...validPayload, notification_frequency: "hourly" });
+    expect(result.success).toBe(false);
+    if (!result.success) {
+      expect(
+        result.error.issues.some((i) => i.path.includes("notification_frequency")),
+      ).toBe(true);
+    }
+  });
+
+  it("rejects when a non-nullable string is null", () => {
+    const result = SettingsSchema.safeParse({ ...validPayload, smtp_host: null });
+    expect(result.success).toBe(false);
+  });
+
+  it("accepts null for nullable secret fields", () => {
+    const result = SettingsSchema.safeParse({ ...validPayload, fireworks_api_key: null });
+    expect(result.success).toBe(true);
+  });
+
+  it("accepts a string for nullable secret fields", () => {
+    const result = SettingsSchema.safeParse({
+      ...validPayload,
+      fireworks_api_key: "fw_test_key",
+    });
+    expect(result.success).toBe(true);
+  });
+
+  it("strips unknown keys (additive backend changes don't break the page)", () => {
+    const result = SettingsSchema.safeParse({ ...validPayload, _unknown_future_field: 42 });
+    expect(result.success).toBe(true);
+    if (result.success) {
+      expect(
+        (result.data as unknown as Record<string, unknown>)._unknown_future_field,
+      ).toBeUndefined();
+    }
+  });
+});

